clippy_lints/src/clamping_without_clamp.rs

```diff
@@ -0,0 +1,932 @@
+            && let Some(second_bin) = try_binary_op(second_cond)
+            && let Some(input_min_max) = is_clamp_meta_pattern(
+                cx,
+                first_bin,
+                second_bin,
+                maybe_min_max_first,
+                maybe_min_max_second,
+                None
+            )
+            {
+                Some(ClampSuggestion {
+                    params: InputMinMax {
+                        input: init,
+                        min: input_min_max.min,
+                        max: input_min_max.max
+                    },
+                    add_semicolon: true,
+                    span: init.span.to(second_expr.span)
+                })
+            } else {
+                None
+            }
+                });
+        Some(iter)
+    } else {
+        None
+    };
+    ret.into_iter().flatten().collect()
+}
+
+fn simplify_kind<'tcx>(expr: &'tcx Expr<'tcx>) -> &'tcx Expr<'tcx> {
+    match &expr.kind {
+        ExprKind::Block(
+            Block {
+                stmts: &[],
+                expr: Some(e),
+                targeted_by_break: false,
+                rules: BlockCheckMode::DefaultBlock,
+                ..
+            },
+            _,
+        )
+        | ExprKind::DropTemps(e) => simplify_kind(e),
+        _ => expr,
+    }
+}
+
+fn is_clampable<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    ty.is_floating_point() || get_trait_def_id(cx, &paths::ORD).map_or(false, |id| implements_trait(cx, ty, id, &[]))
+}
+
+/// `ExprKind::Binary` but more narrowly typed
+struct BinaryOp<'tcx> {
+    op: BinOpKind,
+    left: &'tcx Expr<'tcx>,
+    right: &'tcx Expr<'tcx>,
+}
+
+fn try_binary_op(e: ExprRef<'_>) -> Option<BinaryOp<'_>> {
+    match &e.kind {
+        ExprKind::Binary(op, left, right) => Some(BinaryOp {
+            op: op.node,
+            left,
+            right,
+        }),
+        _ => None,
+    }
+}
+
+/// The clamp meta pattern is a pattern shared between many (but not all) patterns.
+/// In summary, this pattern consists of two if statements that meet many criteria,
+/// - binary operators that are one of [`>`, `<`, `>=`, `<=`].
+/// - Both binary statements must have a shared argument
+///     - Which can appear on the left or right side of either statement
+///     - The binary operators must define a finite range for the shared argument. To put this in
+///       the terms of Rust `std` library, the following ranges are acceptable
+///         - `Range`
+///         - `RangeInclusive`
+///       And all other range types are not accepted. For the purposes of `clamp` it's irrelevant
+///       whether the range is inclusive or not, the output is the same.
+/// - The result of each if statement must be equal to the argument unique to that if statement. The
+///   result can not be the shared argument in either case.
+#[allow(clippy::too_many_lines)] // This code is already fairly succinct given the complexity.
+fn is_clamp_meta_pattern<'tcx>(
```

Comment:
This feels like it could be simplified a good amount, something like

```rust
if eq_expr_value(first_bin.right, first_expr)
    && eq_expr_value(second_bin.right, second_expr)
    && eq_expr_value(first_bin.left, second_bin.left)
{
    match (first_bin.op, second_bin.op) {
        (BinOpKind::Le | BinOpKind::Lt, BinOpKind::Ge | BinOpKind::Gt) => {
            Some(InputMinMax {
                input: first_bin.left,
                min: first_expr,
                max: second_expr,
            })
        },
        (BinOpKind::Ge | BinOpKind::Gt, BinOpKind::Le | BinOpKind::Lt) => {
            Some(InputMinMax {
                input: first_bin.left,
                min: second_expr,
                max: first_expr,
            })
        },
        _ => None,
    }
}
```

We don't have to catch every single combination, particularly if the uncommon ones make the code substantially more complicated

Reply:
> We don't have to catch every single combination, particularly if the uncommon ones make the code substantially more complicated

Trouble is I have no idea which configurations are uncommon.

Reply:
That's fine, just start with the ones that you personally might write is a good rule of thumb. For this function I'd say just `x > max` + `x < min` and `x < min` + `x > max` would be plenty

Extra cases can always be added later too

Reply:
So I found enough ways to simplify the code that I was able to remove the `#[allow(clippy::too_many_lines)]`, without removing any of the combinations. I believe that counts as resolving this issue because you said 

> particularly if the uncommon ones make the code substantially more complicated

Now I can say they don't.